### PR TITLE
Allow comma delimited array settings to have a space after each entry

### LINKED ISSUE #22297: cluster.routing.allocation.exclude._ip using a space after each comma only uses the first ip
**Elasticsearch version**:
Reproducible on:
 - 5.1.1
 - 2.3.3
and I suspect all versions in between.

**Plugins installed**: []
x-pack (this issue should be reproducible without this plugin installed.)

**JVM version**:
1.8.0_111

**OS version**:
Ubuntu - Linux peter-Inspiron-7520 4.4.0-53-generic #74-Ubuntu SMP Fri Dec 2 15:59:10 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
Also tested on OSX Sierra

**Description of the problem including expected versus actual behavior**:
When allocating shards away from more than one node using `cluster.routing.allocation.exclude._ip`, specifying a space after the comma in the comma delimited list causes only the first ip to be used in the exclusion (though all are listed in cluster settings), all subsequent ips in the list are ignored.

My test setup was a couple of machines, with one machine running two instances of ES on different ports and the other with a single instance of elasticsearch, clustered over a network.
The machine with two instances used the same ip address but different ports and my testing was to exclude the shards from allocating to the machine running the single instance.

The method I used for this was to exclude both an unused ip (ot used by any node in the cluster) and the ip of the single instance host, by listing the unused ip first in the list.

The effect was that the first ip in the list was the only one used for the exclusion if using a space after the comma in the list. Though cluster settings showed the entire string when retrieving the settings again. So it was Elasticsearch reading that string but only using the first ip in the list.
Without using any spaces, all the ips in the list were successfully excluded.

I suggest we either:
 - have Elasticsearch strip out the whitespace after commas
 - or we specify a strict format that does not allow spaces after the commas and reject any calls to the API that send an "invalid" string with spaces after the commas with an error message telling the user why it was rejected
 - and with either of the above options, validate that the ips in the list are valid ipv4 or ipv6 ip addresses and throw an error if they are not valid (if we don't do this already)

The current user experience with this configuration setting is poor and needs to be fixed so that Elasticsearch does not "fail" silently and also behaves predictably.
Documentation should also be improved to reflect detail around how Elasticsearch works after whatever method is chosen to fix the bug.

**Steps to reproduce**:

```
# exclude list using fake first ip and real second ip
#

# first with space after comma, nothing happens since only the first (unused) ip is recognised in the cluster allocation exclusion
PUT _cluster/settings
{
  "transient" : {
    "cluster.routing.allocation.exclude._ip" : "192.168.178.35, 192.168.178.21"
  }
}

# second with the space after the comma removed, which ensures all ips are recognized in the allocation exclusion
PUT _cluster/settings
{
  "transient" : {
    "cluster.routing.allocation.exclude._ip" : "192.168.178.35,192.168.178.21"
  }
}

# reset to no ips to reallocate shards evenly again
PUT _cluster/settings
{
  "transient" : {
    "cluster.routing.allocation.exclude._ip" : ""
  }
}
```

**Provide logs (if relevant)**:
Logs only show:
```
[2016-12-21T16:51:37,778][INFO ][o.e.c.s.ClusterSettings  ] [mac] updating [cluster.routing.allocation.exclude.] from [{}] to [{"_ip":"192.168.178.35, 192.168.178.21"}]
[2016-12-21T16:51:58,775][INFO ][o.e.c.s.ClusterSettings  ] [mac] updating [cluster.routing.allocation.exclude.] from [{"_ip":"192.168.178.35, 192.168.178.21"}] to [{"_ip":"192.168.178.35,192.168.178.21"}]
[2016-12-21T16:53:16,101][INFO ][o.e.c.s.ClusterSettings  ] [mac] updating [cluster.routing.allocation.exclude.] from [{"_ip":"192.168.178.35,192.168.178.21"}] to [{"_ip":""}]
```

### PR DESCRIPTION
**Pull request by abeyad**
Previously, certain settings that could take multiple comma delimited
values would pick up incorrect values for all entries but the first if
each comma separated value was followed by a whitespace character.  For
example, the multi-value "A,B,C" would be correctly parsed as
["A", "B", "C"] but the multi-value "A, B, C" would be incorrectly parsed
as ["A", " B", " C"].

This commit allows a comma separated list to have whitespace characters
after each entry.  The specific settings that were affected by this are:

  cluster.routing.allocation.awareness.attributes
  index.routing.allocation.require.*
  index.routing.allocation.include.*
  index.routing.allocation.exclude.*
  cluster.routing.allocation.require.*
  cluster.routing.allocation.include.*
  cluster.routing.allocation.exclude.*
  http.cors.allow-methods
  http.cors.allow-headers

For the allocation filtering related settings, this commit also provides
validation of each specified entry if the filtering is done by _ip,
_host_ip, or _publish_ip, to ensure that each entry is a valid IP
address.

Closes #22297

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java [hunk 1]
```diff
@@ -36,14 +38,28 @@
         OR
     }
 
+    public static final Consumer<Settings> IP_VALIDATOR = (settings) -> {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java]: can you put some javadocs on this

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 2]
```diff
@@ -973,20 +973,26 @@ public IndexMetaData build() {
                     filledInSyncAllocationIds.put(i, Collections.emptySet());
                 }
             }
-            final Map<String, String> requireMap = INDEX_ROUTING_REQUIRE_GROUP_SETTING.get(settings).getAsMap();
+            final Settings requireGroupSettings = INDEX_ROUTING_REQUIRE_GROUP_SETTING.get(settings);
+            DiscoveryNodeFilters.IP_VALIDATOR.accept(requireGroupSettings);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: why don't you use this as a validator on the actual setting?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 3]
```diff
@@ -973,20 +973,26 @@ public IndexMetaData build() {
...
+            final Map<String, String> requireMap = requireGroupSettings.getAsMap();
             final DiscoveryNodeFilters requireFilters;
             if (requireMap.isEmpty()) {
                 requireFilters = null;
             } else {
                 requireFilters = DiscoveryNodeFilters.buildFromKeyValue(AND, requireMap);
             }
-            Map<String, String> includeMap = INDEX_ROUTING_INCLUDE_GROUP_SETTING.get(settings).getAsMap();
+            final Settings includeGroupSettings = INDEX_ROUTING_INCLUDE_GROUP_SETTING.get(settings);
+            DiscoveryNodeFilters.IP_VALIDATOR.accept(includeGroupSettings);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: see above

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 4]
```diff
@@ -973,20 +973,26 @@ public IndexMetaData build() {
...
+            Map<String, String> includeMap = includeGroupSettings.getAsMap();
             final DiscoveryNodeFilters includeFilters;
             if (includeMap.isEmpty()) {
                 includeFilters = null;
             } else {
                 includeFilters = DiscoveryNodeFilters.buildFromKeyValue(OR, includeMap);
             }
+            final Settings excludeGroupSettings = INDEX_ROUTING_EXCLUDE_GROUP_SETTING.get(settings);
+            DiscoveryNodeFilters.IP_VALIDATOR.accept(excludeGroupSettings);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: see above

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java [hunk 5]
```diff
@@ -78,12 +78,12 @@
...
 
     public static final Setting<String[]> CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING =
-        new Setting<>("cluster.routing.allocation.awareness.attributes", "", Strings::splitStringByCommaToArray , Property.Dynamic,
+        new Setting<>("cluster.routing.allocation.awareness.attributes", "", s -> Strings.tokenizeToStringArray(s, ","), Property.Dynamic,
             Property.NodeScope);
     public static final Setting<Settings> CLUSTER_ROUTING_ALLOCATION_AWARENESS_FORCE_GROUP_SETTING =
         Setting.groupSetting("cluster.routing.allocation.awareness.force.", Property.Dynamic, Property.NodeScope);
 
-    private String[] awarenessAttributes;
+    private volatile String[] awarenessAttributes;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java]: 👍

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java [hunk 6]
```diff
@@ -83,9 +84,9 @@ public FilterAllocationDecider(Settings settings, ClusterSettings clusterSetting
         setClusterRequireFilters(CLUSTER_ROUTING_REQUIRE_GROUP_SETTING.get(settings));
         setClusterExcludeFilters(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING.get(settings));
         setClusterIncludeFilters(CLUSTER_ROUTING_INCLUDE_GROUP_SETTING.get(settings));
-        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_REQUIRE_GROUP_SETTING, this::setClusterRequireFilters);
-        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING, this::setClusterExcludeFilters);
-        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_INCLUDE_GROUP_SETTING, this::setClusterIncludeFilters);
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_REQUIRE_GROUP_SETTING, this::setClusterRequireFilters, IP_VALIDATOR);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java]: hmm use the validator on the actual setting instead?

**Review by s1monw — CHANGES_REQUESTED** [review]
I left some questions Looks good though

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java (hunk 1 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java]: done

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 2 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: thanks for alerting me to this, i didn't realize a validator could be set on group settings until you mentioned it.  I fixed this and all other occurrences below

**Comment by abeyad** [comment]
@s1monw I pushed 1ca2d4771b8d479640bac53ab718952508cc388e to address your feedback

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by abeyad** [comment]
retest this please

**Comment by abeyad** [comment]
@s1monw I pushed 953f016a649d87d7fde178e0eeb7474d214566d4 which also prohibits node attributes to have leading or trailing whitespace, if you can take a quick look
/cc @jasontedor

**Review by jasontedor — APPROVED** [review]
LGTM.

**Review by s1monw — APPROVED** [review]
LGTM 2

**Comment by abeyad** [comment]
5.x commit: c763b1e822fe71968bffbe3d003062d1a83784c8

**Comment by abeyad** [comment]
thanks @s1monw @jasontedor

**Pull request merged**